### DISCUSSION TITLE
Django Rest Framework learnings

### DISCUSSION THREAD

**Discussion by lpatmo** [codebuddies/backend#186] [General]
Feel free to put notes here re: what you learn from:

https://www.django-rest-framework.org/tutorial/quickstart/
https://www.django-rest-framework.org/tutorial/1-serialization/

**Comment by lpatmo** [comment]
Really nice to see token-based pagination supported out of the box 😍 

![image](https://user-images.githubusercontent.com/4512699/93542446-ea5bd780-f90d-11ea-8424-10b69246296d.png)

See: https://engineering.mixmax.com/blog/api-paging-built-the-right-way/

**Reply by BethanyG** [reply]
FYI:  absolutely, positively no expert here, but when I set up DRF, I use the following in settings:

` 'DEFAULT_PAGINATION_CLASS': 'rest_framework.pagination.PageNumberPagination',
    'PAGE_SIZE': 10,`

So I may have inadvertently overridden the token-base pagination with "page number" pagination.  Ooops.  But we can re-configure it to something else, I am sure.

**Reply by lpatmo** [reply]
Oh haha, actually I still see "next" and "previous" in the GET /resources, so you're fine! Also I don't think you overrode it because the tutorial called for me to do the same thing ( ```REST_FRAMEWORK = {
    "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.PageNumberPagination",
    "PAGE_SIZE": 10,
}```) as well. 


... the last thing is I may have confused token-based pagination with the appearance of "next" and "previous" fields, when in fact maybe I should have expected tokens 🤔  

In any case, I'm happy with how pagination is working right now.

**Comment by lpatmo** [comment]
On https://www.django-rest-framework.org/tutorial/1-serialization/, not sure I completely understand the `{'base_template': 'textarea.html'}` flag  in the following example of a serializer: 

```
from rest_framework import serializers
from snippets.models import Snippet, LANGUAGE_CHOICES, STYLE_CHOICES


class SnippetSerializer(serializers.Serializer):
    id = serializers.IntegerField(read_only=True)
    title = serializers.CharField(required=False, allow_blank=True, max_length=100)
    code = serializers.CharField(style={'base_template': 'textarea.html'})
    linenos = serializers.BooleanField(required=False)
    language = serializers.ChoiceField(choices=LANGUAGE_CHOICES, default='python')
    style = serializers.ChoiceField(choices=STYLE_CHOICES, default='friendly')

    def create(self, validated_data):
        """
        Create and return a new `Snippet` instance, given the validated data.
        """
        return Snippet.objects.create(**validated_data)

    def update(self, instance, validated_data):
        """
        Update and return an existing `Snippet` instance, given the validated data.
        """
        instance.title = validated_data.get('title', instance.title)
        instance.code = validated_data.get('code', instance.code)
        instance.linenos = validated_data.get('linenos', instance.linenos)
        instance.language = validated_data.get('language', instance.language)
        instance.style = validated_data.get('style', instance.style)
        instance.save()
        return instance
```

**Reply by lpatmo** [reply]
... will revisit later.

**Reply by BethanyG** [reply]
I think that is referring to the textarea widget in a django template....if I recall correctly (_and I probably don't!_) that quickstart project is about collecting code snippetts and saving them for later??  So `code = serializers.CharField(style={'base_template': 'textarea.html'})` is saying that `code` is going to be its _own_ nested serializer that is a character field that pulls from the `base_templates` textarea widget input.....at least that's my working theory.  LOL.

**Reply by lpatmo** [reply]
Thanks, that makes sense! `textarea.html` appeared out of the blue so... I assume I'm going to get a better understanding of it once I have a better understanding of the thing it's asking us to build XD

**Comment by lpatmo** [comment]
model instance -> python native datatypes -> JSON

Can also serialize querysets instead of model instances by passing a flag: 
```
serializer = SnippetSerializer(Snippet.objects.all(), many=True)
```

**Reply by BethanyG** [reply]
WOOT!  That's going to come in handy when we have to make `tagging` and `active hangout` queries!
